{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import forgi\n",
    "import re\n",
    "from os import path, chdir\n",
    "\n",
    "from subprocess import Popen, PIPE\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, cm\n",
    "from mplotlib_addcolors import plot_rna_addcolors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LRTTableFile = \"Results/Tables/TCS_detection_tables/LRTest_table_and_genome_features.tsv\"\n",
    "EcoliChrFile = \"Reference_sequences/NC_000913.3.fasta\"\n",
    "NormCountsTableFile = \"Results/Tables/Merged_ends_counts/N5E_T_vs_NT_CPM_normalized.tsv.gz\"\n",
    "\n",
    "SubsequencesFastaDir = \"hairpin_cleavage/SubsequencesFasta/\"\n",
    "SubsequencesDBNDir = \"hairpin_cleavage/SubsequencesDBN/\"\n",
    "RNAPicturesDir = \"hairpin_cleavage/Pictures/RNASecondaryStructures/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LRTTable = pd.read_csv(LRTTableFile, sep=\"\\t\")\n",
    "LRTTable.sort_values(by=\"PValue.adj\", ascending=True, inplace=True, ignore_index=True)\n",
    "EcoliChrSeq = SeqIO.read(EcoliChrFile, \"fasta\").seq.transcribe()\n",
    "\n",
    "NormCountsTable = pd.read_csv(NormCountsTableFile, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_adj_seqs(seq, pos, strand, width):\n",
    "    assert (strand in [\"+\", \"-\"]), \"Invalid strand value\"\n",
    "    #should be 0-based coordinates\n",
    "    if strand == \"+\":\n",
    "        seq_slice = seq[pos-width : pos+width]\n",
    "    elif strand == \"-\":\n",
    "        seq_slice = seq[pos-width+1 : pos+width+1].reverse_complement()\n",
    "    return(seq_slice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_colors_array(Table, SeqID, pos, strand, width):\n",
    "    if strand == \"+\":\n",
    "        pos_array = np.arange(pos-width, pos+width)\n",
    "        TableSubset = Table[(Table[\"SeqID\"] == SeqID) & (Table[\"Strand\"] == strand) & (Table[\"Pos\"].isin(pos_array))]\n",
    "        TableSubset_sorted = TableSubset.sort_values(by=\"Pos\", ascending=True, ignore_index=True)\n",
    "        logFC_values = TableSubset_sorted[\"aveLogFC\"].tolist()\n",
    "        \n",
    "    elif strand == \"-\":\n",
    "        pos_array = np.arange(pos-width+1, pos+width+1)\n",
    "        TableSubset = Table[(Table[\"SeqID\"] == SeqID) & (Table[\"Strand\"] == strand) & (Table[\"Pos\"].isin(pos_array))]\n",
    "        TableSubset_sorted = TableSubset.sort_values(by=\"Pos\", ascending=False, ignore_index=True)\n",
    "        logFC_values = TableSubset_sorted[\"aveLogFC\"].tolist()\n",
    "    \n",
    "    cmap = cm.Reds\n",
    "    norm = colors.Normalize(vmin=min(logFC_values), vmax=max(logFC_values))\n",
    "    colors_array = [cmap(norm(x)) for x in logFC_values]\n",
    "    colorbar_settings = cm.ScalarMappable(norm=norm, cmap=cmap)\n",
    "    colorbar_settings.set_array(logFC_values)\n",
    "    \n",
    "    return(colors_array, colorbar_settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GetColorsAnnotatedRNAStructures(Table, Width):\n",
    "    ColorsAnnotatedRNAStructures = list()\n",
    "\n",
    "    for i, row in Table.iterrows():\n",
    "        record_id = f'{row[\"MatchedFeatureGene\"]}_{row[\"Pos\"]}_{row[\"Strand\"]}_'\n",
    "        #extract subsequence; coordinates are 0-based\n",
    "        record_seq = extract_adj_seqs(seq=EcoliChrSeq,\n",
    "                                      pos=row[\"Pos\"]-1,\n",
    "                                      strand=row[\"Strand\"],\n",
    "                                      width=Width)\n",
    "        record = SeqRecord(id=record_id,\n",
    "                         name=record_id,\n",
    "                         description=f'{row[\"MatchedFeatureDescription\"]} fragment', \n",
    "                         seq=record_seq)\n",
    "        #write subsequence to separate fasta file\n",
    "        RNAFragmentsOutputFile = path.join(SubsequencesFastaDir, f\"{record_id}.fasta\")\n",
    "        with open(RNAFragmentsOutputFile, \"w\") as hRNAFragmentsOutputFile:\n",
    "            SeqIO.write(sequences=record, handle=hRNAFragmentsOutputFile, format=\"fasta\")\n",
    "        #run RNAfold program, remove energy info from the output (save it) and write dot-bracket notation to file\n",
    "        DotBracketNotationFile = path.join(SubsequencesDBNDir, f\"{record_id}.dbn\")\n",
    "        energy_pattern = r\"\\((.\\d+.+)\\)\"\n",
    "        energy_value = None \n",
    "        with open(DotBracketNotationFile, \"w\") as hDotBracketNotationFile:\n",
    "            RNAfoldProc = Popen([\"RNAfold\", \"--noPS\", RNAFragmentsOutputFile], stdout=PIPE)\n",
    "            for line in RNAfoldProc.stdout:\n",
    "                energy_match = re.search(energy_pattern, line.decode(\"ascii\"))\n",
    "                if energy_match:\n",
    "                    energy_value = energy_match.group(1)\n",
    "                hDotBracketNotationFile.write(re.sub(energy_pattern, \"\", line.decode(\"ascii\")))\n",
    "        #read DBN file\n",
    "        ForgiRecord = forgi.load_rna(DotBracketNotationFile, allow_many=False)\n",
    "        #calculate colors for the heatmap\n",
    "        #here coordinates are 1-based\n",
    "        colors_array, cb_settings = get_colors_array(Table=NormCountsTable, \n",
    "                                                    SeqID=row[\"SeqID\"], \n",
    "                                                    pos=row[\"Pos\"], \n",
    "                                                    strand=row[\"Strand\"], \n",
    "                                                    width=Width)\n",
    "        #append the collected data to the array\n",
    "        ColorsAnnotatedRNAStructures.append({\"id\" : record_id, \n",
    "                                             \"ForgiRecord\" : ForgiRecord, \n",
    "                                             \"energy\" : energy_value, \n",
    "                                             \"colors\" : colors_array, \n",
    "                                             \"cb_settings\" : cb_settings})\n",
    "    return(ColorsAnnotatedRNAStructures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_rna_wrapper(record, label, outputfilepathprefix):\n",
    "    figure, ax = plt.subplots(figsize=(5, 20))\n",
    "\n",
    "    plot_rna_addcolors(record[\"ForgiRecord\"], \n",
    "                       ax=ax, \n",
    "                       text_kwargs={\"fontweight\":\"black\"}, \n",
    "                       lighten=0, \n",
    "                       backbone_kwargs={\"linewidth\":1}, \n",
    "                       colorlist=record[\"colors\"], \n",
    "                       annotations=None)\n",
    "\n",
    "    cbar = figure.colorbar(record[\"cb_settings\"], fraction = 0.03)\n",
    "    cbar.set_label(\"log$_2$FC\", size=20)\n",
    "    cbar.ax.tick_params(labelsize=20)\n",
    "\n",
    "    ax.set_title(label, weight=\"bold\", horizontalalignment=\"left\", position=(0.01, 0.95))\n",
    "    \n",
    "    figure.savefig(f\"{outputfilepathprefix}_.png\", dpi=100)\n",
    "    figure.savefig(f\"{outputfilepathprefix}_.svg\", dpi=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['font.family'] = 'Arial'\n",
    "\n",
    "mRNAStructures = GetColorsAnnotatedRNAStructures(Table=LRTTable[LRTTable[\"MatchedFeatureType\"] == \"CDS\"].head(10), \n",
    "                                                 Width=40)\n",
    "\n",
    "for record in mRNAStructures:\n",
    "    gene = re.search(r\"(\\w+)_\\d+\", record[\"id\"]).group(1)\n",
    "    \n",
    "    plot_rna_wrapper(record=record, \n",
    "                     label=\"$\\mathit{{%s}}$ transcript fragment\" % gene, \n",
    "                     outputfilepathprefix=path.join(RNAPicturesDir, record[\"id\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rRNAStructure = forgi.load_rna(\"hairpin_cleavage/SubsequencesDBN/URS00000ABFE9_588_651.dbn\", allow_many=False)\n",
    "\n",
    "for i, row in LRTTable[LRTTable[\"MatchedFeatureType\"] == \"rRNA\"].head(7).iterrows():\n",
    "    if row[\"Strand\"] == \"+\":\n",
    "        pos_array = np.arange(row[\"MatchedFeatureStart\"]+587, row[\"MatchedFeatureStart\"]+652)\n",
    "        TableSubset = NormCountsTable[(NormCountsTable[\"SeqID\"] == row[\"SeqID\"]) & \n",
    "                                      (NormCountsTable[\"Strand\"] == row[\"Strand\"]) & \n",
    "                                      (NormCountsTable[\"Pos\"].isin(pos_array))]\n",
    "        TableSubset_sorted = TableSubset.sort_values(by=\"Pos\", ascending=True, ignore_index=True)\n",
    "        logFC_values = TableSubset_sorted[\"aveLogFC\"].tolist()\n",
    "        \n",
    "    elif row[\"Strand\"] == \"-\":\n",
    "        pos_array = np.arange(row[\"MatchedFeatureEnd\"]-650, row[\"MatchedFeatureEnd\"]-586)\n",
    "        TableSubset = NormCountsTable[(NormCountsTable[\"SeqID\"] == row[\"SeqID\"]) & \n",
    "                                      (NormCountsTable[\"Strand\"] == row[\"Strand\"]) & \n",
    "                                      (NormCountsTable[\"Pos\"].isin(pos_array))]\n",
    "        TableSubset_sorted = TableSubset.sort_values(by=\"Pos\", ascending=False, ignore_index=True)\n",
    "        logFC_values = TableSubset_sorted[\"aveLogFC\"].tolist()\n",
    "\n",
    "    cmap = cm.Reds\n",
    "    norm = colors.Normalize(vmin=min(logFC_values), vmax=max(logFC_values))\n",
    "    colors_array = [cmap(norm(x)) for x in logFC_values]\n",
    "    colorbar_settings = cm.ScalarMappable(norm=norm, cmap=cmap)\n",
    "    colorbar_settings.set_array(logFC_values)\n",
    "    \n",
    "    figure, ax = plt.subplots(figsize=(5, 20))\n",
    "    plot_rna_addcolors(rRNAStructure, \n",
    "                       ax=ax,\n",
    "                       annotations=None, \n",
    "                       backbone_kwargs={\"linewidth\":1}, \n",
    "                       colorlist=colors_array)\n",
    "\n",
    "    label = \"$\\mathit{%s}$ 588..651 fragment\" % row[\"MatchedFeatureGene\"]\n",
    "    ax.set_title(label, weight=\"bold\", horizontalalignment=\"left\", position=(0.01, 0.95))\n",
    "    \n",
    "    cbar = figure.colorbar(record[\"cb_settings\"], fraction = 0.03)\n",
    "    cbar.set_label(\"log$_2$FC\", size=20)\n",
    "    cbar.ax.tick_params(labelsize=20)\n",
    "    \n",
    "    filenameprefix = f'16S_rRNA_{row[\"MatchedFeatureGene\"]}_fragment'\n",
    "    figure.savefig(path.join(RNAPicturesDir, f\"{filenameprefix}.svg\"), dpi=100)\n",
    "    figure.savefig(path.join(RNAPicturesDir, f\"{filenameprefix}.png\"), dpi=100)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
